src/vsr/journal.zig

```diff
@@ -1837,24 +1855,31 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
         }
 
         pub fn writing(self: *Self, op: u64, checksum: u128) bool {
+            const slot = self.slot_for_op(op);
             var it = self.writes.iterate();
+            var found: bool = false;
             while (it.next()) |write| {
+                const write_slot = self.slot_for_op(write.message.header.op);
+
                 // It's possible that we might be writing the same op but with a different checksum.
                 // For example, if the op we are writing did not survive the view change and was
                 // replaced by another op. We must therefore do the search primarily on checksum.
                 // However, we compare against the 64-bit op first, since it's a cheap machine word.
                 if (write.message.header.op == op and write.message.header.checksum == checksum) {
                     // If we truly are writing, then the dirty bit must be set:
                     assert(self.dirty.bit(self.slot_for_op(op)));
-                    return true;
+                    found = true;
+                } else if (write_slot.index == slot.index) {
+                    // If the in-progress write of '{op, checksum}' will be overwritten by another
+                    // write to the same slot, writing() must return false.
+                    found = false;
```

Comment:
To check if I understand, here we have the op write in progress, and then we see that a different op is going to overwrite the same slot, so that the caller can then overwrite that different op again?

How do we use `writing()` externally?

Reply:
> To check if I understand, here we have the op write in progress, and then we see that a different op is going to overwrite the same slot, so that the caller can then overwrite that different op again?
Correct.

> How do we use `writing()` externally?

Two locations:

1. `Replica.repair_prepare()`: We skip repair a prepare if we are already writing it (either by starting a write if we have it in the pipeline, or sending off a `request_prepare` otherwise).
2. `Replica.write_prepare()`: Don't start writing the prepare if it is already being written.

So if the write queue looks like: `[A, B]` where `A` and `B` are different versions of the same op, and then the replica hits a view change and the new leader picks `A` again, we _should_ queue up a write to `A`, because our pending write to `A` is going to be overwritten by `B`.

I suppose that when the second `A` calls `write_prepare`, we could even scan through the list and remove `B` (which is still waiting for its range lock), and then just allow the first `A` call to do the actual write, though that seems like it would be hard to get right (write?), and difficult to test.

Reply:
Agreed that we shouldn't try and compress the list. My gut feel too. Thanks for the explanation.